### PR TITLE
[UNDERTOW-1623] Fix the deadlock by invoking lastDataRead at IO thread

### PR DESCRIPTION
**Pull request by fl4via**
Jira: https://issues.redhat.com/browse/UNDERTOW-1623

### CODE REVIEW

**Comment by fl4via** [comment]
This is an alternative fix for #828

**Comment by fl4via** [comment]
This is ok to test

File: core/src/main/java/io/undertow/server/protocol/framed/AbstractFramedChannel.java [hunk 1]
```diff
@@ -388,7 +388,7 @@ public synchronized R receive() throws IOException {
             } else if (read == -1 && !hasData) {
                 forceFree = true;
                 readChannelDone = true;
-                lastDataRead();
+                getIoThread().execute(()->{lastDataRead();});
```
Comment (carterkozak) [inline core/src/main/java/io/undertow/server/protocol/framed/AbstractFramedChannel.java]: We can slightly simplify to use a method reference
```suggestion
                getIoThread().execute(this::lastDataRead);
```

Should Http2Channel declare `lastDataRead` volatile?

Comment (fl4via) [reply core/src/main/java/io/undertow/server/protocol/framed/AbstractFramedChannel.java]: @carterkozak good catch! I'll update the PR

Comment (fl4via) [reply core/src/main/java/io/undertow/server/protocol/framed/AbstractFramedChannel.java]: @carterkozak as a second thought, I'm closing this PR and sticking with the original one. I think that asynchronously invoking lastDataRead from receive might have collateral effects on receive logic. The reason is that receive uses isLastFrameReceived() to decide if we are at last frame and  Http2Channel.isLastFrameReceived returns lastDataRead. So, by doing it asynchronously we risk a receive missing a true isLastFrameReceived.

**Comment by fl4via** [comment]
I'm closing this PR and sticking with #828 . I think that asynchronously invoking lastDataRead from receive might have collateral effects on receive logic. The reason is that receive uses isLastFrameReceived() to decide if we are at last frame and Http2Channel.isLastFrameReceived returns lastDataRead. So, by doing it asynchronously we risk a receive missing a true isLastFrameReceived

File: core/src/main/java/io/undertow/server/protocol/framed/AbstractFramedChannel.java (hunk 1 above)
Comment (carterkozak) [reply core/src/main/java/io/undertow/server/protocol/framed/AbstractFramedChannel.java]: That makes sense, I agree that the synchronized approach is safer. Thank you for fixing the bug!

**Pull request closed without merging**